File: app/src/main/java/com/genonbeta/TrebleShot/dialog/RationalePermissionRequest.java
/*
 * Copyright (C) 2019 Veli Tasalı
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

package com.genonbeta.TrebleShot.dialog;

import android.content.Context;
import android.content.pm.PackageManager;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.core.app.ActivityCompat;
import com.genonbeta.TrebleShot.R;
import com.genonbeta.TrebleShot.app.Activity;
import com.genonbeta.TrebleShot.util.AppUtils;

import static com.genonbeta.TrebleShot.activity.HomeActivity.REQUEST_PERMISSION_ALL;

/**
 * created by: Veli
 * date: 18.11.2017 20:16
 */

public class RationalePermissionRequest extends AlertDialog.Builder
{
    public PermissionRequest mPermissionQueue;

    public RationalePermissionRequest(final Activity activity, @NonNull PermissionRequest permission,
                                      boolean killActivityOtherwise)
    {
        super(activity);

        mPermissionQueue = permission;

        setCancelable(false);
        setTitle(permission.title);
        setMessage(permission.message);

        if (ActivityCompat.shouldShowRequestPermissionRationale(activity, mPermissionQueue.permission))
            setNeutralButton(R.string.butn_settings, (dialogInterface, i) -> AppUtils.startApplicationDetails(activity));

        setPositiveButton(R.string.butn_ask, (dialogInterface, i) -> ActivityCompat.requestPermissions(activity,
                new String[]{mPermissionQueue.permission}, REQUEST_PERMISSION_ALL));

        if (killActivityOtherwise)
            setNegativeButton(R.string.butn_reject, (dialogInterface, i) -> activity.finish());
        else
            setNegativeButton(R.string.butn_close, null);
    }

    public static AlertDialog requestIfNecessary(Activity activity, PermissionRequest permissionQueue,
                                                 boolean killActivityOtherwise)
    {
        return ActivityCompat.checkSelfPermission(activity, permissionQueue.permission)
                == PackageManager.PERMISSION_GRANTED ? null : new RationalePermissionRequest(activity, permissionQueue,
                killActivityOtherwise).show();
    }

    public static class PermissionRequest
    {
        public String permission;
        public String title;
        public String message;
        public boolean required;

        public PermissionRequest(String permission, String title, String message)
        {
            this(permission, title, message, true);
        }

        public PermissionRequest(String permission, String title, String message, boolean required)
        {
            this.permission = permission;
            this.title = title;
            this.message = message;
            this.required = required;
        }

        public PermissionRequest(Context context, String permission, int titleRes, int messageRes)
        {
            this(permission, context.getString(titleRes), context.getString(messageRes));
        }
    }
}

File: app/src/main/java/org/monora/uprotocol/client/android/dialog/RationalePermissionRequest.kt
/*
 * Copyright (C) 2019 Veli Tasalı
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package org.monora.uprotocol.client.android.dialog

import android.app.Activity
import android.content.Context
import android.content.DialogInterface
import android.content.pm.PackageManager
import androidx.appcompat.app.AlertDialog
import androidx.core.app.ActivityCompat
import org.monora.uprotocol.client.android.R
import org.monora.uprotocol.client.android.activity.HomeActivity
import org.monora.uprotocol.client.android.util.AppUtils

/**
 * created by: Veli
 * date: 18.11.2017 20:16
 */
class RationalePermissionRequest(
    activity: Activity,
    var permissionQueue: PermissionRequest,
    killActivityOtherwise: Boolean
) : AlertDialog.Builder(activity) {

    data class PermissionRequest(
        val permission: String,
        val title: String,
        val message: String,
        val required: Boolean = true
    ) {
        constructor(context: Context, permission: String, titleRes: Int, messageRes: Int) : this(
            permission,
            context.getString(titleRes),
            context.getString(messageRes)
        )
    }

    companion object {
        fun requestIfNecessary(
            activity: Activity,
            permissionQueue: PermissionRequest,
            killActivityOtherwise: Boolean
        ): AlertDialog? = if (ActivityCompat.checkSelfPermission(activity, permissionQueue.permission)
            == PackageManager.PERMISSION_GRANTED
        ) null
        else
            RationalePermissionRequest(activity, permissionQueue, killActivityOtherwise).show()
    }

    init {
        setCancelable(false)
        setTitle(permissionQueue.title)
        setMessage(permissionQueue.message)
        if (ActivityCompat.shouldShowRequestPermissionRationale(
                activity,
                permissionQueue.permission
            )
        ) setNeutralButton(R.string.butn_settings) { dialogInterface: DialogInterface?, _: Int ->
            AppUtils.startApplicationDetails(activity)
        }
        setPositiveButton(R.string.butn_ask) { _: DialogInterface?, _: Int ->
            ActivityCompat.requestPermissions(
                activity,
                arrayOf(permissionQueue.permission),
                HomeActivity.REQUEST_PERMISSION_ALL
            )
        }
        if (killActivityOtherwise)
            setNegativeButton(R.string.butn_reject) { _: DialogInterface?, _: Int -> activity.finish() }
        else
            setNegativeButton(R.string.butn_close, null)
    }
}